blockdag/ghostdag.go

```diff
@@ -0,0 +1,156 @@
+package blockdag
+
+import "github.com/pkg/errors"
+
+func (dag *BlockDAG) selectedParentAnticone(node *blockNode) (*blockHeap, error) {
+	anticoneSet := newSet()
+	anticoneHeap := newUpHeap()
+	selectedParentPast := newSet()
+	var queue []*blockNode
+	for _, parent := range node.parents {
+		if parent == node.selectedParent {
+			continue
+		}
+		anticoneSet.add(parent)
+		anticoneHeap.Push(parent)
+		queue = append(queue, parent)
+	}
+	for len(queue) > 0 {
+		var current *blockNode
+		current, queue = queue[0], queue[1:]
+		for _, parent := range current.parents {
+			if anticoneSet.contains(parent) || selectedParentPast.contains(parent) {
+				continue
+			}
+			isAncestorOfSelectedParent, err := dag.isAncestorOf(parent, node.selectedParent)
+			if err != nil {
+				return nil, err
+			}
+			if isAncestorOfSelectedParent {
+				selectedParentPast.add(parent)
+				continue
+			}
+			anticoneSet.add(parent)
+			anticoneHeap.Push(parent)
+			queue = append(queue, parent)
+		}
+	}
+	return &anticoneHeap, nil
+}
+
+// blueAnticoneSize returns the blue anticone size of 'block' from the worldview of 'context'.
+// Expects 'block' to be ∈ blue-set(context)
+func (dag *BlockDAG) blueAnticoneSize(block, context *blockNode) (uint32, error) {
+	for current := context; current != nil; current = current.selectedParent {
+		if blueAnticoneSize, ok := current.bluesAnticoneSizes[*block.hash]; ok {
+			return blueAnticoneSize, nil
+		}
+	}
+	return 0, errors.Errorf("block %s is not in blue-set of %s", block.hash, context.hash)
+}
+
+// ghostdag updates newNode.blues, newNode.selectedParent
+// and newNode.bluesAnticoneSizes according to the GHOSTDAG
+// protocol.
+// It updates newNode.blues by going over the anticone of
+// newNode.selectedParent (which is the parent with the
+// highest blue score) and adds it to newNode.blues if it
+// passes two conditions:
```

Comment:
I'd be more concise about "if it passes two conditions":
It's not that the block should pass both conditions, it's that the DAG should not violate these two conditions if the block in question is added to newNode.blues.